### ISSUE TITLE
Proxy objects returned from readonly list methods aren't mutable in change blocks

### ISSUE THREAD

**Issue by airhorns** [automerge/automerge#174]
I am pretty sure that this is a thing Automerge should support: during a change block, iterate a list to find an object in particular and then mutate it. If that iteration is done using something like `.filter`, the objects in the returned list are no longer mutable. 

This test passes works by using a built in list iterator :

```typescript
      it('should allow mutation of objects returned from built in list iteration', () => {
        root = Automerge.change(root, doc => {
          doc.objects = [{foo: 'bar'}, {foo: 'baz'}]
        })
        assert.deepEqual(root.objects, [{foo: 'bar'}, {foo: 'baz'}])

        root = Automerge.change(root, doc => {
          for (let object of doc.objects) {
            if (object.foo == 'baz') {
              object.foo = 'qux'
            }
          }
        })

        assert.deepEqual(root.objects, [{foo: 'bar'}, {foo: 'qux'}])
      })
```

which is as expected, but this test which uses a readonly list method fails:

```typescript
      it('should allow mutation of objects returned from readonly list methods', () => {
        root = Automerge.change(root, doc => {
          doc.objects = [{foo: 'bar'}, {foo: 'baz'}]
        })
        assert.deepEqual(root.objects, [{foo: 'bar'}, {foo: 'baz'}])

        root = Automerge.change(root, doc => {
          // use a proxied iterator function to get new proxies to objects in the list
          const testObjects = doc.objects.filter(object => object.foo == 'baz')
          assert.equal(testObjects.length, 1)
          testObjects[0].foo = "qux"
        })

        assert.deepEqual(root.objects, [{foo: 'bar'}, {foo: 'qux'}])
      })
    })
```

Output:

```
1) Automerge proxy API
       list object
         should support standard read-only methods
           should allow mutation of objects returned from readonly methods:

      AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:

[
  {
    foo: 'bar'
  },
  {
    foo: 'baz'
  }
]

should equal

[
  {
    foo: 'bar'
  },
  {
    foo: 'qux'
  }
]
      + expected - actual

         {
           "foo": "bar"
         }
         {
      -    "foo": "baz"
      +    "foo": "qux"
         }
       ]

      at Context.it (test/proxies_test.js:398:16)
      at processImmediate (internal/timers.js:443:21)
```

This seems like a bug to me as it is diverging from normal JS semantics because of the way the proxy layer works. I poked around in it and I don't understand why the list methods that are just delegated to the JS implementation underneath are returning readonly objects, so any pointers would be appreciated. 

Both of these tests can be found here if needed https://github.com/automerge/automerge/compare/master...airhorns:proxied-list-objects-readonly

**Comment by airhorns** [comment]
Anyone have any pointers they could give me for where and why this might be happening? Happy to try to take a crack at it but I'd appreciate any insight or some tracer bullets I could fire.

**Comment by ept** [comment]
I have changed the behaviour in #187 to use proxy objects, which should enable the pattern you're using. Can you give it a try?

A note on your test cases: in Automerge 0.11.0 both of your tests appear to pass, because as of #179 the objects are no longer immutable by default. But the mutation of the object returned by `.filter()` is an untracked mutation of the Automerge object, not reflected in the actual Automerge state. Enabling the `{freeze: true}` option makes the second test case fail, as it should.

**Comment by ept** [comment]
I have merged #187, which fixes this issue.
